File: examples/int-linked-list/int_linked_list.rs
use std::mem;

pub struct IntegerLinkedList {
    head: Option<Node>,
    num_nodes: usize,
}

struct Node {
    next: Option<Box<Node>>,
    element: i32
}

impl IntegerLinkedList {
    pub fn new() -> IntegerLinkedList {
        IntegerLinkedList { head: None, num_nodes: 0 }
    }

    pub fn size(&self) -> usize {
        self.num_nodes
    }

    pub fn insert(&mut self, index: usize, element: i32) {
        if index == 0 && self.num_nodes > 0 {
            let old_head = mem::replace(&mut self.head, None);
            let new_head =
                Node::new_with_tail(element, Some(Box::new(old_head.unwrap())));
            self.head = Some(new_head);
            self.num_nodes += 1;
            return
        }

        match self.head {
            Some(ref mut head) => {
                head.insert(index, element);
            },
            None => {
                if index == 0 {
                    self.head = Some(Node::new(element));
                } else {
                    panic!("list index out of bounds")
                }
            }
        }
        self.num_nodes += 1
    }

    pub fn remove(&mut self, index: usize) -> i32 {
        let val;
        match self.head {
            Some(ref mut head) => {
                val = head.remove(index);
            },
            None => panic!("list index out of bounds")
        }
        self.num_nodes -= 1;
        val
    }

    pub fn head(&self) -> i32 {
        match self.head {
            Some(ref head) => {
                head.element
            },
            None => panic!("no head of list")
        }
    }

    pub fn tail(&self) -> i32 {
        self.get(self.num_nodes - 1)
    }

    pub fn get(&self, index: usize) -> i32 {
        match self.head {
            Some(ref head) => {
                head.get(index)
            },
            None => panic!("list index out of bounds")
        }
    }
}

impl Node {
    pub fn new(element: i32) -> Node {
        Node { next: None, element: element }
    }

    pub fn new_with_tail(element: i32, next: Option<Box<Node>>) -> Node {
        Node { next: next, element: element }
    }

    pub fn insert(&mut self, index: usize, element: i32) {
        match self.next {
            Some(ref mut next) => {
                if index > 0 {
                    return next.insert(index - 1, element);
                }
            },
            None => {}
        };

        let old_node = mem::replace(&mut self.next, None);
        let mut new_node = Box::new(Node::new(element));
        new_node.next = old_node;
        self.next = Some(new_node)
    }

    pub fn remove(&mut self, index: usize) -> i32 {
        let ret;

        match self.next {
            Some(ref mut next) => {
                if index > 0 {
                    return next.remove(index - 1)
                } else {
                    ret = self.element
                }
            }
            None => if index == 0 {
                ret = self.element
            } else {
                panic!("list index out of bounds")
            }
        };

        self.next = match self.next {
            Some(ref mut next) => { mem::replace(&mut next.next, None) },
            None => None
        };

        ret
    }

    pub fn get(&self, index: usize) -> i32 {
        if index == 0 {
            self.element
        } else {
            match self.next {
                Some(ref next) => next.get(index - 1),
                None => panic!("list index out of bounds")
            }
        }
    }
}

File: examples/int-linked-list/int_linked_list.cpp
#include <memory>
#include <stdexcept>

#include "int_linked_list.hpp"

IntLinkedList::Node::Node(int element) : element(element), next(nullptr) {}
IntLinkedList::Node::Node(int element, std::shared_ptr<Node> next) :
    element(element), next(next) {}

std::shared_ptr<IntLinkedList::IntLinkedList::Node>
IntLinkedList::traverse(size_t num_elements) const {
    std::shared_ptr<Node> current (headNode);

    size_t elements_to_go = num_elements;
    if (num_elements > this->size()) {
        throw std::out_of_range("index out of range");
    }

    while (current->next != nullptr) {
        --elements_to_go;
        current = current->next;
    }

    return current;
}

void IntLinkedList::insert(size_t index, int element) {
    auto current = traverse(index);
    current->next = std::make_shared<Node> (element);
    ++num_nodes;
}

int IntLinkedList::remove(size_t index) {
    // Get the element just before
    auto current = traverse(index - 1);

    int value = current->next->element;
    current->next = current->next->next;

    return value;
}

size_t IntLinkedList::size() const {
    return this->num_nodes;
}

int IntLinkedList::get(size_t index) const {
    return traverse(index)->element;
}

int IntLinkedList::head() const {
    return headNode->element;
}

int IntLinkedList::tail() const {
    return traverse(num_nodes - 1)->element;
}
